test/test_elastic_driver.py

```diff
@@ -0,0 +1,351 @@
+            return 0, time.time()
+
+        driver.start(np=2, create_worker_fn=exec_command)
+        res = driver.get_results()
+        driver.stop()
+
+        assert len(res) == 2
+        for name, (exit_code, timestamp) in res.items():
+            assert exit_code == 0, name
+
+        assert len(rank_results) == 2
+        for rank, (slot_info, updated_slot_info) in rank_results.items():
+            assert updated_slot_info.size == 2, rank
+            assert updated_slot_info.rank == slot_info.rank % 2, rank
+            assert updated_slot_info.local_size == slot_info.local_size, rank
+            assert updated_slot_info.local_rank == slot_info.local_rank, rank
+            assert updated_slot_info.cross_size == 1, rank
+            assert updated_slot_info.cross_rank == 0, rank
+
+    @mock.patch('horovod.run.elastic.driver.DISCOVER_HOSTS_FREQUENCY_SECS', 0.01)
+    def test_rank_and_size_with_host_added(self):
+        """Tests training starts with one host two slots, then a second host is added."""
+        slots = {'host-1': 2}
+        discovery = FixedHosts(slots)
+
+        def add_host():
+            slots = {'host-1': 2, 'host-2': 2}
+            discovery.set(slots)
+
+        driver = ElasticDriver(mock.Mock(), discovery, min_np=2, max_np=4)
+        driver.wait_for_available_hosts(min_np=2)
+
+        rank_results = {}
+
+        def exec_command(slot_info, events):
+            driver.record_ready(slot_info.hostname, slot_info.local_rank)
+
+            if slot_info.hostname == 'host-1':
+                if slot_info.rank == 0:
+                    add_host()
+                driver.wait_for_available_hosts(4)
+                driver.record_ready(slot_info.hostname, slot_info.local_rank)
+
+            driver.record_ready(slot_info.hostname, slot_info.local_rank)
+            updated_slot_info = driver.get_slot_info(slot_info.hostname, slot_info.local_rank)
+            rank_results[slot_info.rank] = (slot_info, updated_slot_info)
+            return 0, time.time()
+
+        driver.start(np=2, create_worker_fn=exec_command)
+        res = driver.get_results()
+        driver.stop()
+
+        assert len(res) == 4
+        for name, (exit_code, timestamp) in res.items():
+            assert exit_code == 0, name
+
+        assert len(rank_results) == 4
+        for rank, (slot_info, updated_slot_info) in rank_results.items():
+            assert updated_slot_info.size == 4, rank
+            assert updated_slot_info.rank == slot_info.rank, rank
+            assert updated_slot_info.local_size == slot_info.local_size, rank
+            assert updated_slot_info.local_rank == slot_info.local_rank, rank
+            assert updated_slot_info.cross_size == 2, rank
+            assert updated_slot_info.cross_rank == slot_info.cross_rank, rank
+
+    @mock.patch('horovod.run.elastic.driver.DISCOVER_HOSTS_FREQUENCY_SECS', 0.01)
+    def test_wait_for_available_hosts(self):
```

Comment:
another test with `min_hosts=2` and `min_np=2` with the same slots sequence as this one to show that discovery waits for the second host even though the first has enough slots